New Pattern: Core Team (#345)
gruetter/innerSourceTest: patterns/2-structured/core-team.md

## Title

Core Team

## Patlet

> Is your InnerSource project difficult for people to work with, yet no one is interested in improving it?

Reviewer: I like the use of a question to open the patlet to invite people who might answer yet to read further. My only comment on this part is the "yet no one is interested in improving it" which seems to imply that even though the IS project is difficult to work with, no one is interested in improving it. It seems to me that this is a causal relationship: because the IS project is difficult to work with, no one is interested in investing the effort to improve it. 
The use of *yet* would need to explore something non-obvious; or else the antecedent would need to focus on why they *should* be interested. E.g., Is your IS project widely needed, yet no one is interested in improving it?

Reviewer 2: I just checked and it looks like we don't have prior instances of a question being used in the Patlet. Maybe because it is a bit harder to express the Problem-Solution combo in a short manner that way?

Anyways, I am all for trying something new :)

Author: Updated language, which was easiest for me to express without a question.

> Establish a core team to take care of the project's fundamental items so that contributors are enabled to add and use the features that provide value to their scenarios.

Reviewer: It's not really emphasized here, but is this a somehow-dedicated core team? Or at least a team whose members have a prioritized emphasis on this IS project's technical health and maintenance? Doesn't need to be necessarily answered here, but probably should be explored a little in the solution part of the write-up.

Author: Updated to include the work `dedicated` in introducing the core team.

## Problem

* It is difficult to contribute to the project.
This could be due to things like:
  * Can't run the project locally.
  * Poor documentation.
  * Convoluted code.
  * Inadequate testing.
* It is difficult to use the project.
Some possible causes:
  * Poor documentation (again).
  * Frequent bugs.
  * Unintuitive setup.

## Story

There's a central project that everyone depends on.
What a great candidate for InnerSource!
Unfortunately, the project has grown organically, with various contributions and additions slapped on haphazardly.

> Now it's an icky, black morass of code that no one understands and everyone is afraid to touch.

Reviewer 2: > an icky, black morass of code 
nice one! 🤣

Author: Thanks!  Although upon re-reading, I chose an adjective other than a color to describe the difficult of working with the code.

> It's clearly due for an overhaul, but even though everyone needs that and wants it, no one takes the time to work on it.

Reviewer 2: If you well to fully spell out the "needs that and wants it" here, what would you say?

Author: Took a stab at more descriptive terms.

## Context

* Many teams use the project.
* Significant tech debt.

> * Slow adoption and iteration on the project.

Reviewer: This (slow adoption) seems to contradict (many teams use the project). Maybe what is meant here is the focus on few contributions (slow iteration)?

Author: Updated the other line to

 > Many teams ~use~ _need_ the project.

> 

Reviewer: I think part of the context is that there is no  core team responsible for the project OR there may be a project owner, but they are not taking responsibility for organizing and promoting contributions to reduce technical debt, etc.

Reviewer 2: I agree. 

If what @NewMexicoKid is describing is indeed the case, then I would expect that to see repeated in multiple places e.g. Patlet + Problem + Forces

Author: Added a bullet point for this feedback, although I'm not sure if it will meet the gist of what you're trying to share?

Reviewer 2: I still have some understanding issues with this patterns. It feels like there are some unspoken assumptions.
The following questions are meant to uncover those:

1. If somebody had asked "who owns/maintains this project", before the Core Team solution was in place, what would the answer have be?
2. Is this Pattern the same as saying "every project should have a maintainer (maintaining team)"?
3. What are the alternatives to get out of this situation, other than a Core Team?

Author: 1.  I think this is in the **Context** now.  "Nobody".
2.  It is one way to move out of the given **Context**.  If someone has this **Context** and wants to move out of it then they can apply this pattern.  If they don't have this **Context** or don't want to move out of it then they don't have to.
3.  I don't know.  If someone had an alternative then they could write another pattern with the same **Context** and a different **Solution**?

Reviewer: Sebastian, for me the context is pretty clear and doesn't need to be specified further. It could match situations where literally no one group or person owns the project (e.g., perhaps someone left the company or changed roles) or where there are nominal owners but not effectively driving the projects anymore (though perhaps that might take a different pattern like a "Friendly Takeover").
I think that a Core Team is probably a good recommendation/pattern for any IS project that fulfills a wide need by many other teams and that isn't well organized/maintained (i.e., matches the context). It is possible there are alternate patterns/solutions to this context, but that isn't a problem for this pattern to solve (though if one later springs up, it could be mentioned as an alternative).

Reviewer 2: Just to give more context about my questions above:

I didn't mean to suggest that this should not be a pattern. It absolutely should be a pattern!
I was just asking these questions because I wanted to confirm why the project doesn't have an owner/maintainer already, given that it is "widely needed".

I have suggested a modification to L38 that would make this more clear to me.

Is possible that in addition to "widely needed" this project is actually already "widely used" as well, i.e. many teams already depend on it? That would also explain the mentioned contributors that only fix their own needs.

I am further suspecting that this project had an owner/maintainer/core team in the past, which got the project to a state where it could become "widely used". However that owner/maintainer/core team somehow vanished?

Some of this extra info could go into the **Story**, or possibly into the **Rationale**.
@NewMexicoKid is making good points above how a project may end up in an "un-owned" and "hard-to-contribute-to" situation, and I suspect many InnerSource practitioners will be able to related to this situation, as it happens everywhere. Still wouldn't hurt to share some examples of why/how this may happen.

Author: There are a few ways that I've seen or could imagine having this **Context** come about.

1.  A need pops up, and a project self-organizes around those needs without ever being officially sponsored.  Over time it's clear that more support is needed.
2.  Someone leaves the company.  They had been supporting an important project out of their side passion and now it is slipping through the cracks since the management structure had never formally recognized and supported the project.
3.  The team that ostensibly owns the project has drifted attention and now is underfunded or focused on other things to the point where they don't give the project the care and feeding that it needs to be successful.

I'm not inclined to put this into the official pattern because I don't think it really matters _how_ you got to the current point.  If you are at this point, then here is a pattern that you can try.  I wouldn't want someone to think this pattern isn't for them just because they way they got to the current **Context** is not what I wrote in the **Story**.

I did update Line 38 if that helps?

Reviewer 2: The updated Line 38 is easier to understand for me. Nice!

Your extra exploration about how a project may end up in the given Context actually really helps me to understand this pattern. So just for me, in my role as a reader of patterns, it is unfortunately that we cannot get that info into the pattern somewhere (doesn't have to be in Context).

However you have your reasons to not add this to the pattern, so let's just go with that approach.

## Forces

* Everyone is busy.
* Each contributing team prioritizes work that results in an immediate payoff for that team.
* As the project grows the natural trend is for it to become more difficult to use and to modify.

## Solution

> Form a core team whose job it is to maintain the project in a state so that others can easily onboard and contribute to it.

Reviewer 2: Is this the "core team" for this project, or "The Core Team" for the whole company?
i.e. is the Core Team taking responsibility for "getting projects to a contribution-ready state" for many projects, or just for one?

Author: This pattern addressees just a single project ... although I thought that from the other writing in the pattern that this point would have been clear already.  Is that not so?

Reviewer 2: My question was related to my confusion about the lack of a maintainer for this project in the first place, which we are discussing in https://github.com/InnerSourceCommons/InnerSourcePatterns/pull/345/files#r691609475. We can close this thread here.

> Form a core team whose job it is to maintain the project in a state so that others can easily onboard and contribute to it.

Reviewer: This might be a good place to indicate what *whose job it is* means in terms of prioritization. I.e., are managers telling specific people that it is now part of their day job to be responsible for organizing and cleaning up the project? Are any members of the core team dedicated in some fashion (say 50% or more) to this project?

Author: Added a paragraph further down with some guidance on actually forming the core team.

This core team does the work that is necessary for a healthy usage and contribution ecosystem.
This critical work tends to not be prioritized as a contribution.
Catogories of this type of work include communication, local environment, and devops infrastructure.

Here are some specific examples:

* Production bugs.
* Documentation.
* Onboarding tutorials and examples.
* Automated testing.
* CI/CD.
* Local environment.
* Modularization.
* Versioning.
* Monitoring.
* Trailblazing new classes/categories of features.

Each of these items is very important to a healthy product ecosystem, yet is unlikely to be prioritized as a contribution.

> Due to their central role, core team members should nearly always fill the role of [Trusted Committers] as well.
> While the [Trusted Committer] role focuses mostly on facilitating others' contribution and use of the project, a core team member regularly contributes to the project as well.

Reviewer 2: ```suggestion
Due to their central role, core team members should nearly always fill the role of [Trusted Committers][tc] as well.
While the [Trusted Committer][tc] role focuses mostly on facilitating others' contribution and use of the project, a core team member regularly contributes to the project as well.
```

I would add an explicit reference here, so that we only have to define it once at the very end.

Author: Good idea.

The core team doesn't have its own business agenda that determines its contibutions.

> They decide what to contribute based on what will most help others to use and contribute to the project.

Reviewer 2: ```suggestion
They decide what to contribute based on what will help others most to use and contribute to the project.
```

I am not sure if this is still proper English but this way around this sentence is easier to understand for me.
By all means feel free to reject this if you don't like it.

Author: Works for me!

A good way to continually remind the core team of this goal is to have them report regularly on:

* number of active teams using the project

> * number of off-team contributions to the project.

Reviewer: suggest: number and scope of off-team contributions to the project.

since some contributions may be major even if fewer in number (not just a numbers game in terms of impact)

Author: I'd like to leave it out in order to keep this guidance simpler and thus easier to implement.

Reviewer 2: ```suggestion
* number of off-team contributions to the project
```

> Continual focus on these metrics will naturally drive the core team to prioritize generally the right work to create a thriving InnerSource ecosystem around the project.

Reviewer: Are these metrics only shared with management that is prioritizing the participation of the core team? Or are these public metrics that can also help indicate to (potential) contributors that this is a healthy project worthy of contributing to?

Also, perhaps a smaller core team can help drive contributions in the areas of reducing technical debt instead of just doing this work themselves?

Author: The most important aspect of the metrics is that the core team itself knows what they are and feels accountable to them.  The mechanism of accountability is less important.

I have observed success of this pattern without sharing these metrics to potential contributors, although it doesn't seem such sharing would hurt.

The core team may choose to care for technical debt by driving contributions in this area.  One way or another, it is responsible for ensuring that debt that inhibits contributions is taken care of.  Either means of doing so (contribution vs. direct coding) can bring that end result.

## Resulting Context

* Many teams use and contribute to the project.

> * It is easy to use and contribute to the project.

Reviewer: I would put this bullet first since it leads other teams to use and contribute to the project.

Author: Good idea.  Done!

* Core team has their success defined in terms of others' interaction with and response to their project.

## Rationale

Separating out a core team and tasking them in this way helps to fill the gaps that a successful project needs yet are left behind by contributors that are pursuing their own agenda only.
The core team fills those gaps and greases the wheels so that the contribution ecosystem remains healthy.

## Known Instances

Nike implented this pattern to manage the InnerSource effort around its reusable CI/CD pipelines.

## Status

Structured

## Author

[Russell R. Rutledge](https://github.com/rrrutledge)

> [Trusted Committers]: https://innersourcecommons.org/learn/learning-path/trusted-committer/
> [Trusted Committer]: https://innersourcecommons.org/learn/learning-path/trusted-committer/

Reviewer 2: ```suggestion
[tc]: https://innersourcecommons.org/learn/learning-path/trusted-committer/
```

## Title

> Core Team

Reviewer 2: When toying with names for this team and pattern, did you have any alternative names in mind?
I am thinking "Contribution Enablement Team" or so.

Especially the following sentence seems essential for the role that this team plays:

> The core team doesn't have its own business agenda that determines its contributions.

Author: I didn't have any others in mind.  I've called it "Core Team" for years, and in my experience it's been a term that others have understood quickly, so I've been happy with it.

This team enables usage as well, so `Contribution Enablement` would miss part of the picture.

Reviewer 2: Ok. I was exploring how in contrast to the "Core Team" a maintaining team would be called **that does have their own business agenda**.

I am happy to go with Core Team, and see how that resonates with other InnerSource practitioners.

## Patlet

> Even when an InnerSource project is widely needed, contributions and usage may be stymied because the project is difficult to work with.

Reviewer 2: Is there a simpler word for "stymied"? Had to look it up.

Author: `hindered`

Establish a core team that is dedicated to take care of the project's fundamental items.
Their work enables contributors to add and use the features that provide value to their scenarios.

## Problem

* It is difficult to contribute to the project.
This could be due to things like:
  * Can't run the project locally.
  * Poor documentation.
  * Convoluted code.
  * Inadequate testing.
* It is difficult to use the project.
Some possible causes:
  * Poor documentation (again).
  * Frequent bugs.
  * Unintuitive setup.

## Story

There's a central project that everyone depends on.
What a great candidate for InnerSource!
Unfortunately, the project has grown organically, with various contributions and additions slapped on haphazardly.
Now it's an icky, thick morass of code that no one understands and everyone is afraid to touch.

> It's clearly due for an overhaul (e.g. refactoring, testing, documentation, etc.), but even though everyone needs and wants that work, no one takes the time to do it.

Reviewer 2: ```suggestion
It's clearly due for an overhaul (e.g. refactoring, testing, documentation, etc.), but even though everyone needs and wants that work to happen, no one takes the time to do it.
```

Author: Looks good - thank you!

## Context

* Many teams need the project.

> * Significant tech debt.

Reviewer 2: ```suggestion
* The project has significant tech debt.
```

* Slow adoption and iteration on the project.

> * Nobody is taking reponsibility for the project and contribution ecosystem as a whole.

Reviewer 2: ```suggestion
* Nobody is taking responsibility for the project and contribution ecosystem as a whole (i.e. the project has no clearly stated owner/maintainer).
```

- one typo fix
- added a clause that mentions owner/maintainer, as those words were the words that I was looking for when trying to understand the Context/Forces

Author: Incorporated this feedback.  What do you think?

## Forces

* Everyone is busy.

> * Each contributing team prioritizes work that results in an immediate payoff for that team.

Reviewer 2: ```suggestion
* Every contributing team is busy, and therefore prioritizes work that results in an immediate payoff for themselves.
```

The "Everyone is busy" seems too generic to stand on its own. Therefore I am proposing to merge these two.

Author: Sure.

Reviewer 2: Taking that as confirmation, hence merging this in.

* As the project grows the natural trend is for it to become more difficult to use and to modify.

## Solution

Form a core team whose job it is to maintain this project in a state so that others can easily onboard and contribute to it.
This core team does the work that is necessary for a healthy usage and contribution ecosystem.
This critical work tends to not be prioritized as a contribution.

> Catogories of this type of work include communication, local environment, and devops infrastructure.

Reviewer 2: ```suggestion
Categories of this type of work include communication, local environment, and DevOps infrastructure.
```

Here are some specific examples:

* Production bugs.
* Documentation.
* Onboarding tutorials and examples.
* Automated testing.
* CI/CD.
* Local environment.
* Modularization.
* Versioning.
* Monitoring.

> * Trailblazing new classes/categories of features.

Reviewer 2: Pure formatting question:
Is there any particular reason to have periods after each list item here, even though those are not full sentences?
Same question for the list in the Problem section.

Author: In the **Problem** section the bullets are finishing the sentence that is started before the colon.

Author: I removed them from this list.

Each of these items is very important to a healthy product ecosystem, yet is unlikely to be prioritized as a contribution.

The core team may be composed of a small number of people on a full-time or a part-time basis.

> The choice depends on the amount of work needed, the availibility of resources, and the culture of the organization.

Reviewer 2: ```suggestion
The choice depends on the amount of work needed, the availability of resources, and the culture of the organization.
```

The most important consideration is to form the team in a way that allows the organization to empower and hold them accountable in the same way as any other team.

> Due to their central role, core team members should nearly always fill the role of [Trusted Committers][tc] as well.

Reviewer 2: About the "nearly always" here:
In which situation should a core team member not be TC?
If we don't have an example of this yet, I would suggest to leave this phrase away to make the statement stronger.

Author: There could be some core team members that don't take on aspects of community stewardship and mentoring, instead just helping with coding the core features.

Reviewer 2: Got it.

While the [Trusted Committer][tc] role focuses mostly on facilitating others' contribution and use of the project, a core team member regularly contributes to the project as well.

> The core team doesn't have its own business agenda that determines its contibutions.

Reviewer 2: ```suggestion
The core team doesn't have its own business agenda that determines its contributions.
```

> They decide what to contribute based on what will help others most to use and contribute to the project.

Reviewer 2: > They decide what to **contribute**..."

This "contribute" might get confused with contributions from off-teams.

Would it be fine here to say:
> They decide what to work on ...

Author: Sure!  Updated.

A good way to continually remind the core team of this goal is to have them report regularly on:

* number of active teams using the project
* number of off-team contributions to the project.

Continual focus on these metrics will naturally drive the core team to prioritize generally the right work to create a thriving InnerSource ecosystem around the project.

## Resulting Context

* It is easy to use and contribute to the project.
* Many teams use and contribute to the project.
* Core team has their success defined in terms of others' interaction with and response to their project.

## Rationale

Separating out a core team and tasking them in this way helps to fill the gaps that a successful project needs yet are left behind by contributors that are pursuing their own agenda only.
The core team fills those gaps and greases the wheels so that the contribution ecosystem remains healthy.

## Known Instances

> Nike implented this pattern to manage the InnerSource effort around its reusable CI/CD pipelines.

Reviewer 2: ```suggestion
Nike implemented this pattern to manage the InnerSource effort around its reusable CI/CD pipelines.
```

  * Unintuitive setup.

## Story

There's a central project that everyone depends on.
What a great candidate for InnerSource!
Unfortunately, the project has grown organically, with various contributions and additions slapped on haphazardly.
Now it's an icky, black morass of code that no one understands and everyone is afraid to touch.
It's clearly due for an overhaul, but even though everyone needs that and wants it, no one takes the time to work on it.

## Context

* Many teams use the project.
* Significant tech debt.
* Slow adoption and iteration on the project.

## Forces

* Everyone is busy.
* Each contributing team prioritizes work that results in an immediate payoff for that team.
* As the project grows the natural trend is for it to become more difficult to use and to modify.

## Solution

Form a core team whose job it is to maintain the project in a state so that others can easily onboard and contribute to it.
This core team does the work that is necessary for a healthy usage and contribution ecosystem.
This critical work tends to not be prioritized as a contribution.
Catogories of this type of work include communication, local environment, and devops infrastructure.

> Here are some specific examples:
> * Production bugs.

Reviewer 2: ```suggestion

Here are some specific examples:

* Production bugs.
```

* Documentation.
* Onboarding tutorials and examples.
* Automated testing.
* CI/CD.
* Local environment.
* Modularization.
* Versioning.
* Monitoring.
* Trailblazing new classes/categories of features.

Each of these items is very important to a healthy product ecosystem, yet is unlikely to be prioritized as a contribution.

Due to their central role, core team members should nearly always fill the role of [Trusted Committers] as well.
While the [Trusted Committer] role focuses mostly on facilitating others' contribution and use of the project, a core team member regularly contributes to the project as well.
The core team doesn't have its own business agenda that determines its contibutions.
They decide what to contribute based on what will most help others to use and contribute to the project.

> A good way to continually remind the core team of this goal is to have them report regularly on:
> * number of active teams using the project

Reviewer 2: ```suggestion
A good way to continually remind the core team of this goal is to have them report regularly on:

* number of active teams using the project
```

* number of off-team contributions to the project.

Continual focus on these metrics will naturally drive the core team to prioritize generally the right work to create a thriving InnerSource ecosystem around the project.

## Resulting Context

* Many teams use and contribute to the project.
* It is easy to use and contribute to the project.
* Core team has their success defined in terms of others' interaction with and response to their project.

## Rationale

Separating out a core team and tasking them in this way helps to fill the gaps that a successful project needs yet are left behind by contributors that are pursuing their own agenda only.
The core team fills those gaps and greases the wheels so that the contribution ecosystem remains healthy.

## Known Instances

Nike implented this pattern to manage the InnerSource effort around its reusable CI/CD pipelines.

> ## Author

Reviewer 2: ```suggestion
## Status

Structured

## Author
```

## Solution

> Form a core team whose job it is to maintain this project in a state so that others can easily onboard and contribute to it.This core team does the work that is necessary for a healthy usage and contribution ecosystem.

Reviewer 2: ```suggestion
Form a core team whose job it is to maintain this project in a state so that others can easily onboard and contribute to it.
This core team does the work that is necessary for a healthy usage and contribution ecosystem.
```

@rrrutledge I suspect that this was an accident, as you got one sentence per row everywhere else?
